### DISCUSSION TITLE
UART read inconsitencies on Pico Pi

### DISCUSSION THREAD

**Discussion by dan-pagan** [micropython/micropython#11785] [RP2040 / Pico] [answered]
I am getting strange results reading tag data from an ID 12LA RFID module. Some tags read correctly, but others don't.
(https://www.sparkfun.com/products/11827)

The data output from the module should be 1 start byte, 10 bytes for the tag ID, 2 checksum bytes, then CR, LF and an end byte.

e.g. on other platforms (Python3 on an RPI zero with a CP2102 to interface to the RFID module) I read :
63 102 249 62 250 30 34 30 26 46 46 230 202 214 242 0
but on the Pico Pi I read :
63 102 249 243 166 121 19 230 76 204 153 153 46 121 61 126 0

I'm using the same module and tags, and I have tested several different tags and RFID modules, all with the same consistent result. Some tags read correctly on both setups, others work on other setups, but fail on the Pico.   

I connect using : 
`uart = UART(0, 9600, parity=None, stop=1, bits=8, rx=Pin(17), tx=None, timeout=1)`

What am I missing? Is the serial read format different on RP2040 devices, or am I missing something with the encoding from the RFID module.  My next idea was to try reading the serial data using C on the Pico to see if that works or not.  I'm starting to go round in circles with this now though, so any help is much appreciated. 

Thanks

P.S. I am using the latest stable version of Micropython (currently v1.20.0).  I started using a Pico Lipo 16mb, but in case the problem was with that board I also tested on a Pico Pi with the same results.

**Comment by dan-pagan** [comment]
I wrote a simple test using Arduino and I got the same results. So this issue isn't specific to MicroPython; It seems to be an issue with the RP2040's hardware uart implementation. 

So a follow on question would be if there are any known issues with the hardware uart implementation on the rp2040, or if anyone else has had issues using it?  
My next step if to try PIO uart.  Does anyone know of a MicroPython implementation of pio uart?

**Comment by robert-hh** [comment]
> if there are any known issues with the hardware uart implementation on the rp2040, or if anyone else has had issues using it?

I did a lot of tests with the RP2 UART, and it worked fine. If possible, check the real timing of the bits. 
Reading the data sheet, none of the two examples seems right. The data does not start with STX (02) and does not end with CR (13) and LF (10). Do you connect Pin 8 or Pin 9 of the RFID module to the RP2? According to the data sheet, it must be Pin 8. Pin 9 is inverted data.

**Reply by dan-pagan** [reply]
Thanks for your reply. I am using Pin 8 with the format selection pin 7 tied to ground for 'ascii' mode.  Section 10 of the datasheet (copied below) says to use pin 8 for 'CMOS' when connecting to computer, and to use pin 9 when connecting directly to a Processor with UART.  Which suggests I should be using pin 9 rather than pin 8 to connect this module to an RP2.  But that means I will need to invert the TTL signal wont I?  I was hoping to connect this module to the RP2 without needing extra external hardware :/  

> 10. Connection direct to a computer
> Direct connection to a computer RS232 can be made by connecting Pin8 to a 1k series resistor and
> connecting the other end of the resistor to the computer RS232 input. The mode is called pseudo
> RS232. On a standard D9 socket, connect module Pin8 via the series 1k to pin2 of the D-type.
> Connect the ground to Pin5 on the D-type. Leave the TX pin3 open. See “Useful Information” below
> for free terminal download information.
> Note that a +2.8v rail will result in the data outputs having a lower swing and may not be suitable for all computers.
> 10.1 Connection to a Processor UART
> Direct connection to UART is made by connecting Pin9 to the UART Rx in pin

**Reply by robert-hh** [reply, marked as answer]
Oh sorry. Pin 9 looks like the right one, and you do not need an inverter. Without data, the output of the Pin should have a high level (~3.3V).

**Reply by dan-pagan** [reply]
Thanks again. I will try pin 9 now.  Reading the docs I noticed that MicroPython also supports inverting the read and write pins.  Good to know.

**Reply by robert-hh** [reply]
Invert is supported on the Pico, but not on all ports. So you could either change the pin or use invert.

**Reply by dan-pagan** [reply]
Pin 9 worked, and pin 8 also worked inverted.  Other than pin 8 being inverted, there doesn't seem to be a difference between the two pins. Not when powering the module with 3.3v at least. 

Sorry for wasting your time with this, but thanks again for your help.
